File: scripts/embed_local.ts
import { PGEssay, PGJSON } from "@/types";
import { loadEnvConfig } from "@next/env";
import fs from "fs";
import { Configuration, OpenAIApi } from "openai";
import stringify from "csv-stringify";

loadEnvConfig("");
const openaiApiKey = process.env.AZURE_OPENAI_APIKEY!;
const openaiEndpoint = process.env.AZURE_OPENAI_ENDPOINT!;
const openaiEmbedding = process.env.AZURE_OPENAI_EMBEDDING!;
const openaiVersion = process.env.AZURE_OPENAI_VERSION!;
let base_url = `${openaiEndpoint}openai/deployments/${openaiEmbedding}`;


const generateEmbeddings = async (essays: PGEssay[]) => {

  const configuration = new Configuration({
    basePath: base_url,
    apiKey: openaiApiKey,
  });
  const openai = new OpenAIApi(configuration);

  const csvStream = fs.createWriteStream("scripts/embeddings.csv", { flags: "w" });
  const csvStringifier = stringify({ header: true });
  csvStringifier.pipe(csvStream);

  for (let i = 0; i < essays.length; i++) {
    const section = essays[i];

    for (let j = 0; j < section.chunks.length; j++) {
      const chunk = section.chunks[j];

      const { essay_title, essay_url, essay_date, essay_thanks, content, content_length, content_tokens } = chunk;

      const embeddingResponse = await openai.createEmbedding({
        model: openaiEmbedding,
        input: content
      },
      {
        headers: {
          "api-key": openaiApiKey,
        },
        params: {
          "api-version": openaiVersion,
        },
      });

      const [{ embedding }] = embeddingResponse.data.data;

      await new Promise((resolve) => setTimeout(resolve, 800));

      csvStringifier.write({
        essay_title,
        essay_url,
        essay_date,
        essay_thanks,
        content,
        content_length,
        content_tokens,
        embedding
      });

      console.log("saved", i, j, essay_title, content_tokens);
    }
  }

  csvStringifier.end();
};

(async () => {
  const book: PGJSON = JSON.parse(fs.readFileSync("scripts/pg.json", "utf8"));

  await generateEmbeddings(book.essays);
})();

File: scripts/embed_local.py
import os
import json
import time

from openai import Configuration, OpenAiApi
from csv import writer


openai_api_key = os.getenv("AZURE_OPENAI_APIKEY")
openai_endpoint = os.getenv("AZURE_OPENAI_ENDPOINT")
openai_embedding = os.getenv("AZURE_OPENAI_EMBEDDING")
openai_version = os.getenv("AZURE_OPENAI_VERSION")

base_url = f"{openai_endpoint}openai/deployments/{openai_embedding}"

def generate_embeddings(essays):
    configuration = Configuration(
        base_path=base_url,
        api_key=openai_api_key
    )
    openai = OpenAiApi(configuration=configuration)

    with open("scripts/embeddings.csv", "w") as f:
        writer = csv.writer(f)
        writer.writerow(["essay_title", "essay_url", "essay_date", "essay_thanks", "content", "content_length", "content_tokens", "embedding"])

        for i, section in enumerate(essays):
            for j, chunk in enumerate(section["chunks"]):
                essay_title = chunk["essay_title"]
                essay_url = chunk["essay_url"]
                essay_date = chunk["essay_date"]
                essay_thanks = chunk["essay_thanks"]
                content = chunk["content"]
                content_length = chunk["content_length"]
                content_tokens = chunk["content_tokens"]

                embedding_response = openai.create_embedding(
                    model=openai_embedding,
                    input_=content,
                    options={
                        "api_key": openai_api_key,
                        "parameters": {
                            "api_version": openai_version
                        }
                    }
                )

                embedding = embedding_response["data"]["data"][0]["embedding"]

                writer.writerow([
                    essay_title,
                    essay_url,
                    essay_date,
                    essay_thanks,
                    content,
                    content_length,
                    content_tokens,
                    embedding
                ])


if __name__ == '__main__':
    with open('scripts/pg.json', 'r') as f:
        book = json.load(f)
    
    generate_embeddings(book["essays"])
